### PR TITLE
parse docker daemon digest and set it in image_ref

### PR DESCRIPTION
**Pull request by enoodle**
This will make sure that `image_ref` will contain the image digest if available (from docker-pullable resource if available or and local docker daemon). This will make it easier to compare images by `image_ref`.

### CODE REVIEW

**Comment by simon3z** [comment]
@enoodle please fix travis
cc @agrare

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb [hunk 1]
```diff
@@ -761,24 +762,45 @@ def parse_image_name(image, image_ref)
...
+                     else
+                       "#{hostname}/"
+                     end
+                   end
+        image_ref = "%{prefix}%{registry}%{name}%{digest}" % {
+          :prefix   => ContainerImage::DOCKER_IMAGE_PREFIX,
+          :registry => registry,
           :name     => image_parts[:name],
-          :digest   => image_parts[:digest],
+          :digest   => if digest.blank? then "" else "@#{digest}" end,
```
Comment (agrare) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: You don't need the explicit `""` so you can get rid of one conditional here

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb [hunk 2]
```diff
@@ -748,7 +749,7 @@ def parse_base_item(item)
     def parse_image_name(image, image_ref)
       # parsing using same logic as in docker
       # https://github.com/docker/docker/blob/348f6529b71502b561aa493e250fd5be248da0d5/reference/reference.go#L174
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: now at https://github.com/docker/distribution/blob/master/reference/reference.go.
dunno if anything significant changed, haven't read old nor new.

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb [hunk 3]
```diff
@@ -761,24 +762,45 @@ def parse_image_name(image, image_ref)
...
+        hostname = image_ref_parts[:host] || image_ref_parts[:host2]
+        port = image_ref_parts[:port]
+        digest = image_ref_parts[:digest]
+      else
+        hostname = image_parts[:host] || image_parts[:host2] || image_parts[:localhost]
+        port = image_parts[:port]
+        digest = image_parts[:digest] || (image_ref_parts[:digest] if image_ref_parts)
+      end
+      unless image_ref.start_with?(ContainerImage::DOCKER_PULLABLE_PREFIX)
+        registry = unless hostname.blank?
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: conditional without else feels a bit weird where a value is expected.  perhaps:
```
registry = hostname.present? && (port.present? ? "#{hostname}:#{port}/" : "#{hostname}/")
```

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb [hunk 4]
```diff
@@ -761,24 +762,45 @@ def parse_image_name(image, image_ref)
...
+      if image_ref.start_with?(ContainerImage::DOCKER_PULLABLE_PREFIX)
+        hostname = image_ref_parts[:host] || image_ref_parts[:host2]
+        port = image_ref_parts[:port]
+        digest = image_ref_parts[:digest]
+      else
+        hostname = image_parts[:host] || image_parts[:host2] || image_parts[:localhost]
+        port = image_parts[:port]
+        digest = image_parts[:digest] || (image_ref_parts[:digest] if image_ref_parts)
+      end
+      unless image_ref.start_with?(ContainerImage::DOCKER_PULLABLE_PREFIX)
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: doesn't this just continue the `else` above?

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb [hunk 5]
```diff
@@ -761,24 +762,45 @@ def parse_image_name(image, image_ref)
...
+      image_ref_parts = docker_pullable_re.match(image_ref) || docker_daemon_re.match(image_ref)
+
+      if image_ref.start_with?(ContainerImage::DOCKER_PULLABLE_PREFIX)
+        hostname = image_ref_parts[:host] || image_ref_parts[:host2]
+        port = image_ref_parts[:port]
+        digest = image_ref_parts[:digest]
+      else
+        hostname = image_parts[:host] || image_parts[:host2] || image_parts[:localhost]
+        port = image_parts[:port]
+        digest = image_parts[:digest] || (image_ref_parts[:digest] if image_ref_parts)
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: can `image_ref_parts` still be nil after this change?  can image_ref be neither format?  can it be missing?

nitpick: postfix `if` is a bit weird inside expression, it's usually used for "do or don't" not for value.  
perhaps `(image_ref_parts && image_ref_parts[:digest])` ?

Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: can `image_ref_parts` still be nil after this change?  can image_ref be neither format?  can it be missing?

nitpick: postfix `if` is a bit weird inside expression, it's usually used for "do or don't" not for value.  
perhaps `(image_ref_parts && image_ref_parts[:digest])` ?

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb [hunk 6]
```diff
@@ -761,24 +762,45 @@ def parse_image_name(image, image_ref)
...
+      end
+      unless image_ref.start_with?(ContainerImage::DOCKER_PULLABLE_PREFIX)
+        registry = unless hostname.blank?
+                     unless port.blank?
+                       "#{hostname}:#{port}/"
+                     else
+                       "#{hostname}/"
+                     end
+                   end
+        image_ref = "%{prefix}%{registry}%{name}%{digest}" % {
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: This overwriting of `image_ref` is not new, but it contributes to the whole thing being hard to follow.
I'm not even following what is `image` vs `image_ref` in the first place :-/
*/me goes to read the specs...  OK I see, though I don't fully understand them.*

I suppose I wish we had different names.  either rename the `image_ref` we receive or the image_ref we output?
Not necessarily in this PR.

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb [hunk 7]
```diff
@@ -761,24 +762,45 @@ def parse_image_name(image, image_ref)
...
+          :prefix   => ContainerImage::DOCKER_IMAGE_PREFIX,
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: is change from DOCKER_PULLABLE_PREFIX deliberate?
previous logic was "if starts docker:// construct a docker-pullable://",
new logic is "if starts docker:// construct a docker://".

looking at [specs diff](https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/9/files#diff-4ccbcf88cf528939845aae98df62c736R163) I suppose it's good.

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb (hunk 5 above)
Comment (enoodle) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: :+1:

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb (hunk 6 above)
Comment (enoodle) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: Yes, it is confusing. Do you have ideas for the names?
I am thinking `docker_image_ref` for the input.

**Comment by simon3z** [comment]
@enoodle @agrare @cben what's the status here? Is this fully verified? Are we ready to merge?

Can anyone review/approve this?

File: app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb (hunk 6 above)
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: `docker_image_ref` for input sounds nice.

Q: is the computed image_ref field even necessary in DB?
what if it became a method on ContainerImage assembling the parts as here on demand?
(not saying do it now, trying to understand better)
is there an invariant of the form "after this PR image_ref can always be computed from the other fields"?  Or does it sometimes contain extra information?

Comment (enoodle) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: @cben If the field was accepted from Openshift/Kubernetes and has "docker-pullable://" prefix then we can leave it as is. Other wise we can't have it as "docker-pullable://" because we can't really check if it is. So we will also need a field to know that.

Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb]: Interesting: we already have `ContainerImage#full_name` and `ContainerImage#docker_id` methods that compute things from image's fields, and peek at the prefix.
Head hurts :-)

**Comment by cben** [comment]
Code LGTM.
I still find it hard to build a mental picture of the scenarios here and how `image_ref` is later used...
Erez, can you give some context what user-visible effects this will have?

**Comment by enoodle** [comment]
@cben we are displaying `image_ref` as the "Image ID" field to the user. Also if it starts with "docker-pullable://" then we use it as the `docker_id` which we send to "image-inspector" as the reference to the image.

**Comment by simon3z** [comment]
@cben @agrare can you review/approve?

**Comment by cben** [comment]
LGTM

**Comment by simon3z** [comment]
@enoodle do we have a BZ?

**Comment by simon3z** [comment]
> @enoodle do we have a BZ?

@enoodle please open one ASAP

**Comment by enoodle** [comment]
Added BZ

**Comment by simaishi** [comment]
Euwe backport (to manageiq repo) details:

```
$ git log -1
commit 500a57483a7a0ecd4eea08c562f6daf1fd3f6b66
Author: Federico Simoncelli <federico.simoncelli@gmail.com>
Date:   Fri May 12 11:55:34 2017 -0400

    Merge pull request #9 from enoodle/parse_container_image_digest_from_docker_daemon
    
    parse docker daemon digest and set it in image_ref
    (cherry picked from commit 3688c988234d69b6963450a2e9b49c1c437f1fc1)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1454446
```

**Comment by simaishi** [comment]
Fine backport (to manageiq repo) details:

```
$ git log -1
commit 17b50c572c5584c33e914a3be20fe983009b4f1d
Author: Federico Simoncelli <federico.simoncelli@gmail.com>
Date:   Fri May 12 11:55:34 2017 -0400

    Merge pull request #9 from enoodle/parse_container_image_digest_from_docker_daemon
    
    parse docker daemon digest and set it in image_ref
    (cherry picked from commit 3688c988234d69b6963450a2e9b49c1c437f1fc1)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1454445
```

**Comment by moolitayer** [comment]
won't this cause duplication of image objects since we changed their primary_key?
@enoodle @cben @simon3z

**Comment by simon3z** [comment]
> won't this cause duplication of image objects since we changed their primary_key?
> @enoodle @cben @simon3z

Ping @cben @enoodle
